python/ray/tune/experiment/trial.py

```diff
@@ -898,6 +911,21 @@ def __getstate__(self):
         state["_state_valid"] = False
         state["_default_result_or_future"] = None
 
+        # Save persistent trial checkpoint data (with relative paths)
+        # Upon load, the path should be constructed again relative to the
+        # `logdir`, which might be updated.
+        checkpoint_states = []
+        for checkpoint in self.get_trial_checkpoints():
+            checkpoint_state = checkpoint.__dict__.copy()
+            checkpoint_dir = checkpoint.dir_or_data
+            if isinstance(checkpoint_dir, ray.ObjectRef):
+                checkpoint_dir = ray.get(checkpoint_dir)
```

Comment:
We should never run in this situation. The code is actually quite confusing, but a bit of context:

- On trial checkpoint save we set `trial.saving_to = _TrackedCheckpoint(object_ref, ...)`
- But `trial.saving_to` is not serialized. It's also not tracked by the trial checkpoint manager
- Thus it will not be included in `self.get_trial_checkpoints()`
- Instead, once the future is resolved, it will be stored (as resolved) in `dir_or_data` and only _then_ be tracked form the trial checkpoint manager
- Thus, we should never run into the situation where we have an object ref here

Thus we can actually assert that this is not a object ref here, or just leave it out, or skip if it is. I'd probably prefer an assertion here